snapcraft/internal/sources/_base.py

```diff
@@ -41,6 +41,36 @@ def __init__(self, source, source_dir, source_tag=None, source_commit=None,
         self.source_details = None
 
         self.command = command
+        self._checked = False
+
+    def check(self, target: str):
+        """Check if pulled sources have changed since target was created.
+
+        :param str target: Path to target file.
+        """
+        self._checked = True
+        return self._check(target)
+
+    def update(self):
+        """Update pulled source.
+
+        :raises RuntimeError: If this function is called before `check()`.
+        """
+        if not self._checked:
+            # This is programmer error
+            raise RuntimeError("source must be checked before it's updated")
+        self._update()
+
+    def _check(self, target: str):
```

Comment:
why not just do the abc. thing here, you can leave the docstring, it will still be valid code.

Reply:
Because then we'd need to add it to all sources, even those that don't actually support doing this.